### DISCUSSION TITLE
`add` and `sub` functions

### DISCUSSION THREAD

**Discussion by definim** [PaulRBerg/prb-math#28] [Ideas]
`add` and `sub` have presumably been left out of these (really useful - thanks!) libraries because with solc 0.8 they need do nothing more than `+` and `-`.

I would like to argue for including `add` and `sub` functions in the library, nonetheless.

Using such functions may save devs/reviewers/auditors/idealised-end-users-who-actually-grok-code the mental diversion that is "she's using this library for decimals but just adding them as if they are integers; maybe that's a bug; I'd better make sure it's compatible with the underlying implementation".

Consumers of the library can probably achieve the same clarity by adding code comments, but it feels neater just to expose trivial `add` and `sub` functions to separate these concerns.

Would be happy to PR if the reaction is favourable, but @paulrberg suggested starting a discussion here in the first instance. :)

**Comment by definim** [comment]
Worth noting that if the code _was_ ever changed improve type safety as mooted in https://github.com/hifi-finance/prb-math/discussions/29, then the need for `add` and `sub` functions would become objective.

**Comment by PaulRBerg** [comment]
This has been implemented in [v2.0.0](https://github.com/hifi-finance/prb-math/releases/tag/v2.0.0). I'm locking the conversation.

**Comment by PaulRBerg** [comment]
I have recently released [v3.0.0](https://github.com/paulrberg/prb-math/releases/tag/v3.0.0), and I thought that an update here would be useful for future visitors.

PRBMath is now built with [user defined value types][udvt] and [free functions][ff] instead of the library syntax, and there are various non-mathematical helper functions for these value types:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity >=0.8.13;

import "@prb/math/UD60x18.sol";

function addSub() pure returns (bool result) {
  UD60x18 x = ud(1e18);
  UD60x18 y = ud(2e18);
  UD60x18 y = ud(3e18);
  result = x.add(y).sub(z);
}
```

However, you should note that using these functions instead of the vanilla operators (e.g. + and -) will result in a higher gas cost.

For more detail on this, refer to the [Helper Functions](https://github.com/paulrberg/prb-math/tree/b6066c8794ff2eaa6ccf667612fbe4bddad8a9a3#helper-functions) section in the README.

[udvt]: https://blog.soliditylang.org/2021/09/27/user-defined-value-types/
[ff]: https://twitter.com/PaulRBerg/status/1450444297925431301
